SwiftUI Simple Template Example

langleyd opened the pull request:
See the [issue](https://github.com/vector-im/element-ios/issues/4781) for the list of items we are looking to address in this PR

### What Example 1 (User Profile) looks like
<img width=300 src="https://user-images.githubusercontent.com/6135282/132244131-7c4497f4-7362-4ad0-b4d6-ba57f8508657.png">

closes https://github.com/vector-im/element-ios/issues/4781

langleyd commented on RiotSwiftUI/Modules/Template/SimpleProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,61 @@
+        
+        userService.presencePublisher
+            .map(TemplateProfileStateAction.updatePresence)
+            .receive(on: DispatchQueue.main)
```
The current design is that all dispatches happen on the main thread. This simplifies the implementation and will usually meet the requirements for most ViewModels. As we get into implementing larger/more complex views we will likely need to do this work off them main thread(a list like room bubbles where we may do iteration/filtering etc across collections).

We could move towards an implementation like [mvrx](https://github.com/airbnb/mavericks/tree/master/mvrx/src/main/kotlin/com/airbnb/mvrx) that the android team use. Their 'withState'/`setState` approach does the the work in the background and in a thread safe way. This approach/api may also suit `async`/`await` and we could find good commonality across the platforms with an mvrx type approach to view models and using kotlin coroutins/ swift async/await.

langleyd commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+        
+        userService.presencePublisher
+            .map(TemplateUserProfileStateAction.updatePresence)
+            .receive(on: DispatchQueue.main)
```
The current design is that all dispatches happen on the main thread. This simplifies the implementation and will usually meet the requirements for most ViewModels. As we get into implementing larger/more complex views we will likely need to do this work off them main thread(a list like room bubbles where we may do iteration/filtering etc across collections).

We could move towards an implementation like mvrx that the android team use. Their 'withState'/setState approach does the the work in the background and in a thread safe way. This approach/api may also suit async/await and we could find good commonality across the platforms with an mvrx type approach to view models and using kotlin coroutins/ swift async/await.

langleyd replied:
Again similar to the comment below, if we have the common pieces in a superclass we can make this syntactically nicer and be able to do something like:

```swift
userService.presencePublisher
  .map(TemplateUserProfileStateAction.updatePresence)
  .dispatchActionTo(self)
```

langleyd commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+    
+    /**
+     A redux style reducer, all modifications to state happen here. Recieves a state and a state action and produces a new state.
+     */
```
Usually in swift I've see the `inout` style used for reducers as it's a bit neater. the alternative is:

```swift
  private func reducer(state: TemplateUserProfileViewState, action: TemplateUserProfileStateAction) -> TemplateUserProfileViewState {
        var newState = state
        switch action {
        case .updatePresence(let presence):
            newState.presence = presence
        }
    return newState
    }
```

langleyd replied:
Don't think it will make a difference to performance. Swift doesn't do a copy on assignment, it optimises things under the cover and only does a copy when there is a mutation to the underlying value.

langleyd commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+    // MARK: - Private methods
+    /**
+     Send state actions to mutate the state.
+     */
```
This could be made common. In `mxrx` there is an open ViewModel class to subclass that contains the state store that can be iterated on.

SBiOSoftWhare commented on RiotSwiftUI/Modules/Common/Avatar/Service/MatrixSDK/MXAvatarService.swift:
```diff
@@ -19,12 +19,12 @@ import MatrixSDK
 import DesignKit
 
-enum AvatarServiceError: Error {
+enum MXAvatarServiceError: Error {
```
Prefixes like `MX` are only needed for ObjC compatible libraries. Otherwise we can rely on module name for disambiguation. MatrixSDK.AvatarService.

langleyd replied:
So we this is the Matrix implementation of the protcol within `Riot` target. Would you prefer `MatrixSDKAvatarService`? We also have a mock implementation `MockAvatarService`? and might have `KotlinAvatarService` / `RustAvatarService` in the future.

SBiOSoftWhare replied:
For the default class we can use `AvatarService` for the moment like we do for other classes used in Swift in the Element app. When disambiguation will come we will think about having different targets if needed or different repos. `KotlinAvatarService` will be named `AvatarService.kt` and probably a bridge will be made to make it usable by `AvatarService.swift` so we should not make us some issues.

SBiOSoftWhare commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfile.swift:
```diff
@@ -0,0 +1,68 @@
+import SwiftUI
+
+@available(iOS 14.0, *)
+struct TemplateUserProfile: View {
```
It can be interesting if we can have pragma marks like we have in view controller templates

langleyd replied:
Any chance you could provide an example? Not sure I fully follow.

langleyd replied:
Oh actually maybe I do, let me know if what I do is correct 😂

SBiOSoftWhare commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Model/TemplateUserProfileViewState.swift:
```diff
@@ -0,0 +1,23 @@
+struct TemplateUserProfileViewState {
+    let avatar: AvatarInputType?
+    let displayName: String?
+    var presence: TemplateUserProfilePresence = .offline
```
Having immutable structure with immutable properties is safer to guarantee functional programming principles. We can set the default value like this otherwise:
```
init(avatar: AvatarInputType?, displayName, presence = .offline)
```

SBiOSoftWhare commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Coordinator/TemplateUserProfileCoordinator.swift:
```diff
@@ -0,0 +1,64 @@
+    // MARK: - Setup
+    
+    @available(iOS 14.0, *)
+    init(session: MXSession) {
```
We are now using an input parameters class for Coodinators

langleyd replied:
I saw the PR for that but couldn't see the discussion you referenced before. Any details of that just for context of those changes?

SBiOSoftWhare replied:
Yes we are now using a `struct` as input parameter
```
struct TemplateUserProfileCoordinatorParamers {
 let session: MXSession
}
```
To make init refactoring easier and handle default values and convenience init aside.

SBiOSoftWhare commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Service/MatrixSDK/MXTemplateUserProfileService.swift:
```diff
@@ -0,0 +1,83 @@
+    init(session: MXSession) {
+        self.session = session
+        
+        let listenerReference = session.myUser.listen { [weak self] event in
```
Maybe we can use a lazy var here

langleyd replied:
Not quite sure how you mean?

SBiOSoftWhare replied:
```
private lazy listenerReference: Any = {
buildPresenceListener()
}()
```

langleyd replied:
And then where do we call the new property `listenerReference`?

langleyd replied:
And where would this be called? We'd need to do something this to setup the presence updates in the `init`:
```swift
init() {
  let _ = listenerReference
}
```

Which seems to a bit non-obvious that there are side effects.
I should probably actually handle the nil case in case SDK does not return a listener and log an error if setup fails.
I'll move it out of the init into a function that sets up the listener.
Maybe have a look and let me know if you have any more thoughts.

manuroe commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,89 @@
+        case .updatePresence(let presence):
+            state.presence = presence
+        }
+    }
```
Can we log here to understand how the state is evolving?
Action and new state should be enough.

Our previous suffers for the lack of logs. They are hard to debug from rageshakes. We need to have more logs in the template in the first place.

langleyd replied:
Logger added. Added a log statement as you suggested. Logging in reducers could end up being too high frequency and as they print sometimes large structs can be very large statements. Suppose we can keep an eye on it, but might make it difficult to follow the rest of the the logs.

manuroe reviewed (commented):
Some comments before I switched context

langleyd commented on RiotSwiftUI/Modules/Common/Avatar/Service/MatrixSDK/AvatarService.swift:
```diff
@@ -33,6 +33,10 @@ class AvatarService: AvatarServiceType {
     
     private let mediaManager: MXMediaManager
     
+    static func instantiate(mediaManager: MXMediaManager) -> AvatarServiceProtocol {
```
Well for the dependency injection to work, you need to pass the object to `addDependency` cast to the [type that it is used at injection time.](https://github.com/vector-im/element-ios/blob/develop/RiotSwiftUI/Modules/Common/Avatar/ViewModel/AvatarViewModel.swift#L28).

So we can cast it at the point we call add dependency? In my head it made sense to have this factory method that returned the protocol we use at runtime within the module(AvatarServiceProtocol). Granted should have named it better `makeAbstractService` or something.

langleyd commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,89 @@
+    
+    // MARK: - Private
+    /**
+     Send state actions to mutate the state.
```
How do you mean? include Parameters and Return type etc?

manuroe replied:
yes

langleyd commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,89 @@
+     */
+    private func dispatch(action: TemplateUserProfileStateAction) {
+        var newState = self.viewState
+        reducer(state: &newState, action: action)
```
Well I think this should probably be made common rather than repeated so wouldn't be as concerned.

But yes it can actually. I was thinking that doing it the way you suggested would mean multiple mutations within the reducer would cause the publisher to fire multiple times. But it won't as its a value type and "copied" when passed between functions. So save to pass `viewState` directly./

pixlwave replied:
I'd vote for `reducer(state: &viewState, action: action)` to drop the `self.` too 😉

langleyd commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,89 @@
+    /**
+     A redux style reducer, all modifications to state happen here. Recieves a state and a state action and produces a new state.
+     */
+    private func reducer(state: inout TemplateUserProfileViewState, action: TemplateUserProfileStateAction) {
```
Yea good idea, thanks.

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfile.swift:
```diff
@@ -0,0 +1,76 @@
+        }
+        .frame(maxHeight: .infinity)
+        .navigationTitle(viewModel.viewState.displayName ?? "")
+        .navigationBarItems(leading: leftButton, trailing: rightButton)
```
As we're decided on iOS 14 as our target, please can we switch to using `.toolbar` here instead? It would give  more flexibility 🙂

langleyd replied:
great point

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,89 @@
+    }
+    
+    // MARK: - Public
+    func proccess(viewAction: TemplateUserProfileViewAction) {
```
```suggestion
    func process(viewAction: TemplateUserProfileViewAction) {
```

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfileHeader.swift:
```diff
@@ -0,0 +1,58 @@
+    
+    // MARK: Public
+    
+    @Environment(\.theme) var theme: ThemeSwiftUI
```
Again, probably should be private.

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfilePresenceView.swift:
```diff
@@ -0,0 +1,66 @@
+    // MARK: - Properties
+    
+    // MARK: Public
+    let presense: TemplateUserProfilePresence
```
```suggestion
    let presence: TemplateUserProfilePresence
```

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,89 @@
+    func proccess(viewAction: TemplateUserProfileViewAction) {
+        switch viewAction {
+        case .cancel:
+            self.cancel()
```
```suggestion
            cancel()
```

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift:
```diff
@@ -0,0 +1,89 @@
+        case .cancel:
+            self.cancel()
+        case .done:
+            self.done()
```
```suggestion
            done()
```

pixlwave commented on RiotSwiftUI/Modules/Common/Test/XCTestPublisherExtensions.swift:
```diff
@@ -0,0 +1,55 @@
+ */
+@available(iOS 14.0, *)
+extension XCTestCase {
+    func xcAwait<T: Publisher>(
```
This is a nice addition! I think it would be worthwhile having a small example code block in a doc comment for the function.

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Service/TemplateUserProfileServiceProtocol.swift:
```diff
@@ -0,0 +1,36 @@
+}
+
+@available(iOS 14.0, *)
+extension TemplateUserProfileServiceProtocol {
```
I can see why these are here, but it feels a bit weird.

langleyd replied:
Will have a think

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfile.swift:
```diff
@@ -0,0 +1,76 @@
+            )
+            Divider()
+            VStack{
+                HStack(alignment: .center){
+                    Spacer()
+                    Text("More great user content!")
+                        .font(theme.fonts.title2)
+                        .foregroundColor(theme.colors.secondaryContent)
+                    Spacer()
+                }
```
A simpler option (which I'm trying to re-train my *"I love stacks"* brain to use more often):
```suggestion
                Text("More great user content!")
                    .font(theme.fonts.title2)
                    .foregroundColor(theme.colors.secondaryContent)
                    .frame(maxWidth: .infinity)
```

pixlwave replied:
Turns out you don't even need the `.frame(maxWidth: .infinity)` in this instance as the VStack is centre aligned by default anyway 🙃

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfilePresenceView.swift:
```diff
@@ -0,0 +1,66 @@
+            Image(systemName: "circle.fill")
+                .resizable()
+                .frame(width: 8, height: 8)
+                .foregroundColor(foregroundColor)
```
The foreground colour should be inherited here anyway so this is probably redundant.

langleyd replied:
oh yea, leftover from an edit. Thanks.

pixlwave commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfileHeader.swift:
```diff
@@ -0,0 +1,58 @@
+    var body: some View {
+        VStack {
+            if let avatar = avatar {
+                HStack{
+                    Spacer()
+                    AvatarImage(avatarData: avatar, size: .xxLarge)
+                    Spacer()
+                }
+                .padding(.vertical)
```
Same as above this could become
```suggestion
                AvatarImage(avatarData: avatar, size: .xxLarge)
                    .padding(.vertical)
```

langleyd commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfile.swift:
```diff
@@ -0,0 +1,76 @@
+    
+    // MARK: Public
+    
+    @Environment(\.theme) var theme: ThemeSwiftUI
```
Oh for some reason in my head I thought environmentVariables couldn't be private. I was, in fact, wrong. :)

langleyd commented:
Hey @pixlwave @manuroe  @SBiOSoftWhare et al,

I have added a common ViewModel implementation in  [StateStoreViewModel](RiotSwiftUI/Modules/Common/ViewModel/StateStoreViewModel.swift). Looking at the documentation there is probably a good place to start. Then you can see an implementation of it in [TemplateUserProfileViewModel](RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift) and the use of the viewModel context in the [main screen](RiotSwiftUI/Modules/Template/SimpleUserProfileExample/View/TemplateUserProfile.swift).

The second example which is coming will include an example of two-way binding to state from a textfield(sending a message).

I've also added the idea of screen state which can be used both for rendering example previews and also for writing UI tests that cover the different states. You can see an example screen state in [MockTemplateUserProfileScreenState](RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Service/Mock/MockTemplateUserProfileScreenState.swift)
and the corresponding [UI Tests](RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Test/UI/TemplateUserProfileUITests.swift).

It's also now a 1-liner to do previews of the top level screens, whereas we used to have to initialise services, mocks, VMs` etc inline. Can see it on the [preview of the main screen](https://github.com/vector-im/element-ios/pull/4791/files#diff-6174f2b8f91d242c6f2b4f6aed714e8d35db21aa08eec9412806581d3a24f567R71).

Best way will be to load it up and play with things. But might be difficult to take in all the things until actually going to use the template at some point, when we can continue to iterate on it.

manuroe commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Test/UI/TemplateUserProfileUITests.swift:
```diff
@@ -0,0 +1,53 @@
+    func verifyTemplateUserProfilePresence(presence: TemplateUserProfilePresence) {
+        let presenceText = app.staticTexts["presenceText"]
+        XCTAssert(presenceText.exists)
+        XCTAssert(presenceText.label == presence.title)
```
`XCTAssertEqual` is more future proof for future debug.

manuroe commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Test/UI/TemplateUserProfileUITests.swift:
```diff
@@ -0,0 +1,53 @@
+    func verifyTemplateUserProfileLongName(name: String) {
+        let displayNameText = app.staticTexts["displayNameText"]
+        XCTAssert(displayNameText.exists)
+        XCTAssert(displayNameText.label == name)
```
XCTAssertEqual

manuroe commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Coordinator/TemplateUserProfileCoordinator.swift:
```diff
@@ -0,0 +1,64 @@
+    
+    // MARK: - Public
+    func start() {
+        templateUserProfileViewModel.completion = { [weak self] result in
```
Can we log the start and the completion callback to understand the navigation?

manuroe reviewed (commented):
Is it possible to add a button to demonstrate a view action via the new `Context`thing that (indirectly or not) impacts the model?

manuroe commented on RiotSwiftUI/Modules/Template/SimpleUserProfileExample/Model/TemplateUserProfileStateAction.swift:
```diff
@@ -17,5 +17,7 @@
 import Foundation
 
 enum TemplateUserProfileStateAction {
+    case incrementCount
```
I wonder if we should try to categorise state actions by type (user action, service action)
There would be something like:
```
case viewAction(TemplateUserProfileViewAction)
```

It would avoid to duplicate things. The view model will be aware of all user actions. `done` is probably useless but `cancel` can be useful to cancel any pending request.

langleyd replied:
Yea we can do that for view actions that go directly to state modification we can pass on the view event.

Not sure I fully understand the second bit. Do you mean in the reducer? We should only be doing synchronous state modifications in the reducer.

If we are doing any work in the service that has a subscription stored in `cancellables` they will all be cancelled on `deinit` of the viewmodel.
But yea if there is any other manual cleanup we needed to do we could use `cancel` action in `process(viewAction:)`

langleyd replied:
Updated with the first suggestion you made.

langleyd commented:
Tested for any regressions on the SwiftUI features we already had, looks ok.
